packages/hash/frontend/src/blocks/page/createPlaceholderPlugin/createPlaceholderPlugin.tsx

```diff
@@ -0,0 +1,96 @@
+import { paragraphBlockComponentId } from "@hashintel/hash-shared/blocks";
+import {
+  componentNodeToId,
+  findComponentNodes,
+} from "@hashintel/hash-shared/prosemirror";
+import { Schema } from "prosemirror-model";
+import { Plugin, PluginKey } from "prosemirror-state";
+import { Decoration, DecorationSet } from "prosemirror-view";
+import { RenderPortal } from "../usePortals";
+import { Placeholder } from "./Placeholder";
+
+interface PlaceholderPluginState {
+  focused: boolean;
+}
+
+const placeholderPluginKey = new PluginKey<PlaceholderPluginState, Schema>(
+  "placeholderPlugin",
+);
+
+export const createPlaceholderPlugin = (renderPortal: RenderPortal) => {
+  return new Plugin<PlaceholderPluginState, Schema>({
+    key: placeholderPluginKey,
+    state: {
+      init() {
+        return { focused: false };
+      },
+      apply(tr, state) {
+        const focused = tr.getMeta(placeholderPluginKey);
+
+        // return the old state if transaction does not have focused meta
+        if (focused === undefined) {
+          return state;
+        }
+
+        return { focused };
+      },
+    },
+    props: {
+      decorations(state) {
+        const firstNode = state.selection.$anchor.node(1);
+
+        if (!firstNode) return;
+
+        const componentId = componentNodeToId(
+          findComponentNodes(firstNode)[0]!,
+        );
+
+        const isFocused = placeholderPluginKey.getState(state)?.focused;
+        const isParagraph = componentId === paragraphBlockComponentId;
+        const isEmpty = findComponentNodes(firstNode)[0]?.childCount === 0;
+
+        const showPlaceholder = isParagraph && isEmpty && isFocused;
+
+        if (!showPlaceholder) return;
+
+        const widgetPos = state.selection.$anchor.posAtIndex(0, 1);
+
+        const placeholderDecoration = Decoration.widget(
+          widgetPos,
+          () => {
+            const mountNode = document.createElement("div");
+
+            renderPortal(<Placeholder />, mountNode);
```

Comment:
Does this need a destroy to do `renderPortal(null, mountNode)` somewhere to tell React to forget about it?

Reply:
ProseMirror handles removing the node since we return it from this function AFAIK

Reply:
If Prosemirror removes a node which react is supposed to be rendering into, you get a crash. So you need to unmount the react tree before Prosemirror removes the node.

Reply:
In the video below, ProseMirror removes the node here and we don't get a crash? I am not sure what needs to be changed here. Let me know if I'm missing something.
https://user-images.githubusercontent.com/39553853/186072634-879101e0-f96e-489f-8694-a16d246041e6.mov

Reply:
I think the crashes you can see when you don't inform react that you're going to destroy a node it renders to are hard to reproduce / intermittent, but we already see them in some places and I'm wary of introducing the possibility of more.

You can see what I think you need to add at https://github.com/hashintel/hash/pull/953/files#r952441689

Reply:
I got the point, thanks for the explanation. It's implemented here https://github.com/hashintel/hash/pull/953/commits/72c97874ff1ab83993d30a63dfe5cfc11438b92d